### ISSUE TITLE
Separate MODAK Build and Run Stages

### ISSUE THREAD

**Issue by spresser-hlrs** [SODALITE-EU/application-optimisation#17]
MODAK currently produces jobscripts that have only a single stage. This can waste resources. For example, if a job requires GPUs to run, those GPUs will currently sit idle while the image is built. If this takes significant time, this can be a significant waste of valuable resources.

Alter MODAK’s job scripts to have distinct “build” and “run” stages such that the build stage can be run on a cluster head or login node without tying up the full job resources.

Please let me know if this needs more detail and/or discussion at the technical meeting.

**Comment by dev-zero** [comment]
Could it also be a possibly to run the build step as a separate job without GPU reservation on the cluster instead to avoid the resource constraints and jitter on the login nodes and to get proper accounting information for the build step? Depending on the queue manager one could even set this is up as 2 dependent jobs.

**Comment by spresser-hlrs** [comment]
Yes, it's possible. There are some possible problems I can see. On some of the systems I've been on, the intent of the login nodes is to be used for builds. Running builds on them is free for that reason. Using a job node would cost money and they sometimes have only read-only versions of certain filesystems. Predicting where the output could go could be a challenge. Additionally, some systems prevent job nodes from communicating with the internet (this is an issue at HLRS, for example). However, I'm not sure how common issues like this are in reality.

However, the idea is elegant, especially because it fits in well with what MODAK already does well.

I don't have a strong opinion one way or the other. I think we should ask @alazzaro and @kmlTE  to weigh in -- they may know things we have missed.

**Comment by kmlTE** [comment]
@dev-zero, @spresser-hlrs There is no standard way to do this. It all depends on cluster policies and the size and constraints of the application we want to compile. But typically the login nodes are used for the compilations. At HLRS, some build with `nice` command to respect the resources. The further issues I see when compile as a job is the waiting queue time if the resources are overbooked.

In any cases, we anyhow separate the build and run stages in MODAK. At the end there will be 2 scripts: 1) either a shell script that compiles on login node or a job script that compiles on the compute node and 2) actual application run as job script. Can we be flexible and provide an additional flag for MODAK DSL for building, where we specify where it should compile, e.g. login/compute? What do you think?
```json
"job": {
  "job_options": {
    ...
  },
  "target": {
    ...
  },
  "optimisation": {
    ...
  },
  "application": {
    "app_tag": "test",
    "executable": "python3 main_mpi.py -s -f input",
    "build": {
      "src": "some-repo.git",
      "build_command": "python3 setup.py install --user",
      "build_target": "compute | login"
    }
  }
}
```

**Comment by dev-zero** [comment]
@spresser-hlrs good point about the cost and it is indeed likely that we can most of the time build on login nodes. The only example I've found so far were ARM systems where the login nodes were x86_86 and cross-compilation was impossible for some packages and very cumbersome for others.
@kmlTE as a newcomer to the project I am still in the process of getting acquainted with all the details, but it feels the wrong place to put this information: at the "build spec" level I would prefer not to make too many assumptions about the build environment. While it might make sense to formulate requirements for the build itself like how much memory needs to be available (which might be important for heavily templated C++ projects, or when doing LTO) or something like requesting the same arch as the compute nodes for cases where cross-compilation is broken, explicitly specifying a build target introduces knowledge about the target environment in the build spec.

**Comment by spresser-hlrs** [comment]
I was thinking about this issue more last night. All of the directives for the scheduler are set up so they look like comments to the shell. Therefore, I think we can produce the same file, regardless of whether it will be run in the shell or as a job via the scheduler. And that means that the AADM is the only piece of software that cares where this script runs -- and it only cares because it is doing the running.

So I think that MODAK doesn't care where this script will run. That means we don't have to make a modification to the DSL and that I can build this feature on top of the existing infrastructure for building jobscripts.

Does this make sense? I'm going to go ahead and start (re-)implementing it this way -- I think it gives us the maximum flexibility and speeds up how quickly the feature can be complete.

**Comment by dev-zero** [comment]
@spresser-hlrs Yes, it does. As I wrote above, from my point of view the only info which might be required in the DSL are restrictions which are a-priori known and which MODAK could try to resolve (at some point in the future) depending on the knowledge of  available build environments.
If it is ok for you I'm going to merge #20 as it already fixes part of #18 and introduces some best practices wrt coding style.

**Comment by kmlTE** [comment]
@spresser-hlrs as you said the orchestrator should know what MODAK returns: either a job script or shell script. Do you have in mind a return parameter that allows to determine that?

**Comment by kmlTE** [comment]
@spresser-hlrs I understand the point that they both are valid shell scripts, but the way of how to run them is different. For shell scripts, we can run them directly: `./script.sh`, but for job scripts, the script should be submitted with the scheduler specific command, e.g. `qsub` or `sbatch`. If MODAK returns the build instruction in the same script (shell or job), the orchestrator needs to know how to run it, either it should submit it via scheduler command to the compute node or run as shell on the login node. 

Of course we can introduce the wrapper script that will always be run as a shell script, but inside it will contain either build commands or the scheduler submission command along with the build commands (see below), but for me it seems a bit complicated:

Wrapper for login node:
```bash
#!/bin/bash
# build_commands
cd software
make && make install
EOF
```
Wrapper for compute node:
```bash
#!/bin/bash
qsub <<EOF
#!/bin/bash
#PBS headers if needed
# build_commands
cd software
make && make install
EOF
```

**Comment by spresser-hlrs** [comment]
Oh, I see! I misunderstood what you were asking.

Is there a reason the wrapper has to come from MODAK? As I was imagining it, I thought the orchestrator would invoke the resulting script however the user configured it to via the AADM. This could be a simple option that can be either "run via job manager" or "run on login node". Based on that, the orchestrator could invoke the script. It could add and use the wrapper script if necessary or could directly call the script, qsub, or sbatch command. Is there a reason this is more technically complex or doesn't work?

**Comment by kmlTE** [comment]
@spresser-hlrs nope, that should be okay and this is similar to my previous comment with specifying build target, except it won't be the MODAK DSL, but rather AADM/RM property that specifies that. I totally agree with that. However, now I have the following question, how will MODAK know about what job parameters are needed to generate the job script for building? I don't think we should encode job script generation into AADM/RM, since MODAK is already doing that for generation of job script that runs the application and AADM/RM rely on that. And those job parameters for building are different from the job parameter for running the application, e.g. different number of nodes, memory, etc.

**Comment by spresser-hlrs** [comment]
True. At the moment, I am creating a new request for the jobscript generator to process. It includes:

- The unmodified `application` and `target` sections from the original request.
- A modified version of the `job_options` section. This is limited to one node and one process and renames the job and output files by prefixing them with "build-"
- All other sections are removed.

This lets me produce a build file that uses a minimum of resources. For example, for this request:

```
{
  "job": {
    "job_options": {
      "job_name": "test",
      "wall_time_limit": "1:00:00 ",
      "node_count": 1,
      "process_count_per_node": 2,
      "standard_output_file": "test.out",
      "standard_error_file": "test.err",
      "combine_stdout_stderr": true,
      "copy_environment": true
    },
    "application": {
      "app_tag": "test",
      "executable": "bash -c \"cd modpy/build/lib/moduli/ && python3 main_mpi.py -s -f $PBS_O_WORKDIR/output/iso/L1L2-iso.dens -o $PBS_O_WORKDIR/output/moduli\"",
      "build": {
        "src": "$HOME/modpy/.git",
        "build_command": "bash -c \"cd modpy && pip3 --no-cache-dir install numpy pandas mpi4py --user && python3 setup.py install --user\""
      }
    },
    "target": {
      "name": "egi"
    },
    "optimisation": {
      "enable_opt_build": true,
      "enable_autotuning":true,
      "app_type":"hpc",
      "opt_build": {
        "cpu_type": "x86",
        "acc_type": ""
      },
      "autotuning": {
        "tuner": "cresta",
        "input": "dsl text"
      },
      "app_type-hpc": {
        "config":{
          "parallelisation":"mpi"
        },
        "data":{
          "location": "L1L2_NonLinear_prepared.tar.gz"
        },
        "parallelisation-mpi": {
          "library": "openmpi",
          "version": "3.1.3"
        }
      }
    }
  }
}
```

It produces the following build script:
```
#!/bin/bash
## START OF HEADER ##
#SBATCH -J test_build
#SBATCH -N 1
#SBATCH --ntasks-per-node=1
#SBATCH --output=build-test.out
#SBATCH --error=build-test.err
#SBATCH --export=ALL 
## END OF HEADER ##
cd $SLURM_SUBMIT_DIR
export PATH=$SLURM_SUBMIT_DIR:$PATH
 git clone $HOME/modpy/.git
bash -c "cd modpy && pip3 --no-cache-dir install numpy pandas mpi4py --user && python3 setup.py install --user"
```

I'm still testing this against other scenarios, but this build script looks like it does what we've been discussing. Does it look reasonable to you?

**Comment by spresser-hlrs** [comment]
- I'll wait for additional input from @alazzaro and @dev-zero , but I think we could easily add an optional parameter for build parallelism? Alternatively, we could look for options in the build command, but that seems extremely fragile.
- That's correct, but I am not currently handling the optimized container runtime case. I will add that.

**Comment by dev-zero** [comment]
>     * I'll wait for additional input from @alazzaro and @dev-zero , but I think we could easily add an optional parameter for build parallelism? Alternatively, we could look for options in the build command, but that seems extremely fragile.

Wrt build parallelism: I agree, it needs to be defined separately, not parsed from a command. IMHO the easiest way to apply it is by variable substitution in the build command. Ex. `build_command: make -j {{NUMJOBS}}` (or similar).

**Comment by spresser-hlrs** [comment]
I've put up a PR for this - #29 . It includes a BUILD_PARALLELISM option and ensures it uses the optimized container runtime.

**Comment by dev-zero** [comment]
@spresser-hlrs thanks, looks good, review added. Let's discuss at the meeting today on how to proceed with merging wrt to PR #28. For Jinja2 dep: I was in the process of preparing a PR with most of the JobGenerator replaced with Jinja2, hoping it will give a clearer picture on the final script (and allowing for easier addition of new scheduler/wrapper scripts).
